crates/polars-stream/src/nodes/io_sources/multi_file_reader/extra_ops/apply.rs

```diff
@@ -0,0 +1,250 @@
+                    // the extra columns were attached by the reader (which is just Parquet when it has a predicate).
+                    incoming_schema.clone()
+                } else {
+                    // We use a trick to determine our schema state before reordering by applying onto an empty DataFrame.
+                    // This is much less error prone compared determining it separately.
+                    let mut df = DataFrame::empty_with_schema(incoming_schema);
+                    slf.apply_to_df(&mut df, IdxSize::MAX)?;
+                    df.schema().clone()
+                };
+
+                if cfg!(debug_assertions)
+                    && schema_before_reorder.len() != final_output_schema.len()
+                {
+                    assert_eq!(schema_before_reorder, final_output_schema);
+                    unreachable!()
+                }
+
+                let initialized_reorder =
+                    ReorderColumns::initialize(&final_output_schema, &schema_before_reorder);
+
+                let Self::Initialized { reorder, .. } = &mut slf else {
+                    unreachable!()
+                };
+
+                *reorder = initialized_reorder;
+
+                // Return a `Noop` if our initialized state does not have any operations. Downstream
+                // can see the `Noop` and avoid running through an extra distributor pipeline.
+                let slf = match slf {
+                    Initialized {
+                        row_index: None,
+                        cast_columns: None,
+                        extra_columns,
+                        predicate: None,
+                        reorder: ReorderColumns::Passthrough,
+                    } if extra_columns.is_empty() => Self::Noop,
+
+                    Initialized { .. } => slf,
+
+                    _ => unreachable!(),
+                };
+
+                Ok(slf)
+            },
+        }
+    }
+
+    /// # Panics
+    /// Panics if `self` is `Uninitialized`
+    pub fn apply_to_df(
+        &self,
+        df: &mut DataFrame,
+        current_row_position: IdxSize,
+    ) -> PolarsResult<()> {
+        let Self::Initialized {
+            row_index,
+            cast_columns,
+            extra_columns,
+            predicate,
+            reorder,
+        } = ({
+            use ApplyExtraOps::*;
+
+            match self {
+                Noop => return Ok(()),
+                Uninitialized { .. } => panic!("ApplyExtraOps not initialized"),
+                Initialized { .. } => self,
+            }
+        })
+        else {
+            unreachable!();
+        };
+
+        if let Some(ri) = row_index {
+            unsafe {
+                df.with_column_unchecked(Column::new_row_index(
+                    ri.name.clone(),
+                    ri.offset.saturating_add(current_row_position),
+                    df.height(),
+                )?)
+            };
+            df.clear_schema();
```

Comment:
Why is this clearing the schema? Either `with_column_unchecked` should do that or it wouldn't need to happen at all.